micro_sam/training/joint_sam_trainer.py

```diff
@@ -0,0 +1,214 @@
+import os
+import time
+from typing import Optional, Union
+
+import torch
+import torch.nn as nn
+from torchvision.utils import make_grid
+
+from .sam_trainer import SamTrainer
+
+from torch_em.model import UNETR
+from torch_em.trainer.logger_base import TorchEmLogger
+from torch_em.model.unet import Decoder, ConvBlock2d, Upsampler2d
+
+
+class CustomDecoder(Decoder):
+    "To make use of the `V-Net` level logic - as we can't make use of the skip connections"
+    def forward(self, x):
+        for block, sampler in zip(self.blocks, self.samplers):
+            x = sampler(x)
+            x = block(x)
+
+        return x
+
+
+class UNETRForJointTraining(UNETR):
+    def __init__(
+        self,
+        encoder: Optional[nn.Module] = None,
+        out_channels: int = 1,
+        final_activation: Optional[Union[str, nn.Module]] = None,
+        **kwargs
+    ) -> None:
+        super().__init__(encoder, out_channels, **kwargs)
+
+        self.encoder = encoder
+
+        # parameters for the decoder network
+        depth = 3
+        initial_features = 64
+        gain = 2
+        features_decoder = [initial_features * gain ** i for i in range(depth + 1)][::-1]
+        scale_factors = depth * [2]
+        self.out_channels = out_channels
+
+        self.decoder = Decoder(
+            features=features_decoder,
+            scale_factors=scale_factors[::-1],
+            conv_block_impl=ConvBlock2d,
+            sampler_impl=Upsampler2d
+        )
+
+        self.out_conv = nn.Conv2d(features_decoder[-1], out_channels, 1)
+        self.final_activation = self._get_activation(final_activation)
+
+    def forward(self, x):
+        org_shape = x.shape[-2:]
+
+        x = torch.stack([self.preprocess(e) for e in x], dim=0)
+
+        x = self.encoder(x)
+        x = self.decoder(x)
+
+        x = self.out_conv(x)
+        if self.final_activation is not None:
+            x = self.final_activation(x)
+
+        x = self.postprocess_masks(x, org_shape, org_shape)
+
+        return x
+
+
+class JointSamTrainer(SamTrainer):
+    def __init__(
+            self, **kwargs
+    ):
+        super().__init__(**kwargs)
+        self.unetr = UNETRForJointTraining(
+            img_size=self.model.img_size,
+            backbone="sam",
+            encoder="vit_b",
```

Comment:
I think we need to extract the encoder from the sam model and then pass it here.